java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
+        }
+    }
+
+    private suspend fun <T : ActualParcelable<U>, U, V> runWithResult(
+        intent: Intent,
+        parcelable: KClass<T>, // dummy argument for type inference
+        block: suspend (U) -> V
+    ) = runWithResult(intent) {
+        val argument = intent.getParcelableExtra<T>(OPERATION_ARG)?.actual
+        argument?.let { block(it) }
+    }
+
+    /**
+     * Execute an [ArcHost] method in the block, encode return into a [Bundle] and send via
+     * [ResultReceiver]
+     */
+    private suspend fun runWithResult(
+        intent: Intent, block: suspend () -> Any?
+    ) {
+        val bundle = Bundle()
+        val result = block()
+
+        when (result) {
+            is List<*> -> {
+                val arrayList = ArrayList<ParcelableParticleIdentifier>()
+                result.forEach {
+                    arrayList += (it as ParticleIdentifier).toParcelable()
+                }
+                bundle.putParcelableArrayList(
+                    OPERATION_RESULT, arrayList
+                )
+            }
+            is Parcelable -> bundle.putParcelable(
+                OPERATION_RESULT, result
+            )
+            is Operation -> bundle.putInt(
+                OPERATION_RESULT, result.ordinal
+            )
+            else -> Unit
+        }
+        // This triggers a suspended coroutine to resume with the value.
+        intent.getParcelableExtra<ResultReceiver>(
+            OPERATION_RECEIVER
+        )?.send(0, bundle)
+    }
+
+    internal enum class Operation {
+        START_ARC,
+        STOP_ARC,
+        REGISTER_PARTICLE,
+        UNREGISTER_PARTICLE,
+        GET_REGISTERED_PARTICLES
+
+    }
+
+    companion object {
+        const val ARC_HOST_INTENT = "arcs.android.host.ARC_HOST"
+        private const val OPERATION = "OPERATION"
+        private const val OPERATION_ARG = "OPERATION_ARG"
+        private const val OPERATION_RECEIVER = "OPERATION_RECEIVER"
+        private const val OPERATION_RESULT = "OPERATION_RESULT"
+
+        internal fun createArcHostIntent(
+            operation: Operation,
+            component: ComponentName,
+            argument: Parcelable?
+        ): Intent = Intent(
+            ARC_HOST_INTENT
+        ).setComponent(component)
+            .putExtra(OPERATION, operation.ordinal)
+            .putExtra(
+                OPERATION_ARG, argument
+            )
+
+        @VisibleForTesting
+        fun onResult(intent: Intent, receiver: ResultReceiver?) =
+            intent.putExtra(
+                OPERATION_RECEIVER, receiver
+            )
+
+        @VisibleForTesting
+        fun getParticleIdentifierListResult(resultData: Bundle?): List<ParticleIdentifier>? =
+            resultData?.getParcelableArrayList<ParcelableParticleIdentifier>(
+                OPERATION_RESULT
+            )?.map { it -> it.actual }
```

Comment:
Might be more usable to return an empty list rather than a null list if `resultData` is null